Library/Homebrew/os/linux/elf.rb

```diff
@@ -149,7 +149,10 @@ def needed_libraries_using_patchelf_rb(path)
     end
 
     def find_full_lib_path(basename)
-      local_paths = (path.patchelf_patcher.runpath || path.patchelf_patcher.rpath)&.split(":")
+      local_paths = (path.patchelf_patcher.runpath || path.patchelf_patcher.rpath)&.split(":")&.map do |rp|
+        # Expand instances of $ORIGIN in the R(UN)PATH
+        rp.gsub "$ORIGIN", path.parent
```

Comment:
```suggestion
      local_paths = (path.patchelf_patcher.runpath || path.patchelf_patcher.rpath)&.split(":")&.map do |runpath|
        runpath.gsub "$ORIGIN", path.parent
```

Reply:
You could move this `#gsub` call into the `local_paths&.each` block below, I think.

Also: I'm not sure if we want `#gsub`, since `$ORIGIN` is only replaced by the dynamic linker if it is the prefix of the path, IIUC.

Reply:
> I'm not sure if we want `#gsub`, since `$ORIGIN` is only replaced by the dynamic linker if it is the prefix of the path, IIUC.

`$ORIGIN` gets substituted even if it's not the prefix; check [this](https://github.com/bminor/glibc/blob/glibc-2.35/elf/dl-load.c#L287-L354). Or TL;DR: a quick example:

```console
linuxbrew@9c46a48f740f:~/.linuxbrew/opt/python/bin$ patchelf --print-rpath python3.12
/home/linuxbrew/.linuxbrew/Cellar/python@3.12/3.12.3/lib
linuxbrew@9c46a48f740f:~/.linuxbrew/opt/python/bin$ ldd python3.12
	linux-vdso.so.1 (0x00007fff28bb6000)
	libpython3.12.so.1.0 => /home/linuxbrew/.linuxbrew/Cellar/python@3.12/3.12.3/lib/libpython3.12.so.1.0 (0x00007b09b6838000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007b09b660c000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007b09b6523000)
	/home/linuxbrew/.linuxbrew/lib/ld.so => /lib64/ld-linux-x86-64.so.2 (0x00007b09b6eb9000)
linuxbrew@9c46a48f740f:~/.linuxbrew/opt/python/bin$ patchelf --set-rpath '$ORIGIN/../lib' python3.12
linuxbrew@9c46a48f740f:~/.linuxbrew/opt/python/bin$ ldd python3.12
	linux-vdso.so.1 (0x00007ffcba9d2000)
	libpython3.12.so.1.0 => /home/linuxbrew/.linuxbrew/Cellar/python@3.12/3.12.3/bin/./../lib/libpython3.12.so.1.0 (0x0000701941109000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x0000701940edd000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x0000701940df4000)
	/home/linuxbrew/.linuxbrew/lib/ld.so => /lib64/ld-linux-x86-64.so.2 (0x000070194178a000)
linuxbrew@9c46a48f740f:~/.linuxbrew/opt/python/bin$ patchelf --set-rpath '/usr/bin/../../$ORIGIN/../lib' python3.12
linuxbrew@9c46a48f740f:~/.linuxbrew/opt/python/bin$ ldd python3.12
	linux-vdso.so.1 (0x00007ffc6119c000)
	libpython3.12.so.1.0 => /usr/bin/../..//home/linuxbrew/.linuxbrew/Cellar/python@3.12/3.12.3/bin/./../lib/libpython3.12.so.1.0 (0x00007cd2a2506000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007cd2a22da000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007cd2a21f1000)
	/home/linuxbrew/.linuxbrew/lib/ld.so => /lib64/ld-linux-x86-64.so.2 (0x00007cd2a2b8a000)
```

Also, as can be seen in the path expansion logic, two more such tokens exist: `$PLATFORM` and `$LIB`. See https://www.akkadia.org/drepper/dsohowto.pdf, page 40, paragraph "`$ORIGIN` is not the only DST available...". Though, I think they are less commonly seen and can be followed up later.

Reply:
Moved the expansion into the `local_paths&.each` block that follows, and replaced `#gsub` with just `#sub` (and only matches `$ORIGIN` from beginning of string for each path in the list).

Reply:
Ah, oops. Will switch it back to a `gsub` for anywhere in the string (and will also add that link as a comment).